src/main/java/com/amazon/opendistroforelasticsearch/security/resolver/IndexResolverReplacer.java

```diff
@@ -280,112 +333,41 @@ public boolean replace(final TransportRequest request, boolean retainMode, Strin
-            if (this == o) return true;
-            if (o == null || getClass() != o.getClass()) return false;
-            IndexResolveKey that = (IndexResolveKey) o;
-            return enableCrossClusterResolution == that.enableCrossClusterResolution &&
-                    opts.equals(that.opts) &&
-                    Arrays.equals(original, that.original);
-        }
-
-        @Override
-        public int hashCode() {
-            return Boolean.hashCode(enableCrossClusterResolution) + 31 * opts.hashCode() + 127*Arrays.hashCode(original);
-        }
-    }
-
     public Resolved resolveRequest(final Object request) {
         if(log.isDebugEnabled()) {
             log.debug("Resolve aliases, indices and types from {}", request.getClass().getSimpleName());
         }
 
-        final Resolved.Builder resolvedBuilder = new Resolved.Builder();
-        final AtomicBoolean isIndicesRequest = new AtomicBoolean();
-        // set of previously resolved index requests to avoid resolving
-        // the same index more than once while processing bulk requests
-        final Set<IndexResolveKey> alreadyResolved = new HashSet<>();
-
-        getOrReplaceAllIndices(request, (original, localRequest, supportsReplace) -> {
-            final IndicesOptions indicesOptions = indicesOptionsFrom(localRequest);
-            final boolean enableCrossClusterResolution = localRequest instanceof FieldCapabilitiesRequest || localRequest instanceof SearchRequest;
-            final IndexResolveKey key = new IndexResolveKey(indicesOptions, enableCrossClusterResolution, original);
-            // skip the whole thing if we have seen this exact resolveIndexPatterns request
-            if (!alreadyResolved.contains(key)) {
-                final Resolved iResolved = resolveIndexPatterns(key.opts, key.enableCrossClusterResolution, key.original);
-                alreadyResolved.add(key);
-                resolvedBuilder.add(iResolved);
-                isIndicesRequest.set(true);
-
-                if (log.isTraceEnabled()) {
-                    log.trace("Resolved patterns {} for {} ({}) to {}", original, localRequest.getClass().getSimpleName(), request.getClass().getSimpleName(), iResolved);
-                }
-            }
-            return IndicesProvider.NOOP;
-        }, false);
-
-        if(!isIndicesRequest.get()) {
-            //not an indices request
-            return Resolved._LOCAL_ALL;
-        }
+        final ResolvedIndicesProvider resolvedIndicesProvider = new ResolvedIndicesProvider(request);
 
-        if(log.isTraceEnabled()) {
-            log.trace("Finally resolved for {}: {}", request.getClass().getSimpleName(), resolvedBuilder.build());
-        }
+        getOrReplaceAllIndices(request, resolvedIndicesProvider, false);
 
-        return resolvedBuilder.build();
+        return resolvedIndicesProvider.resolved();
     }
 
-    public final static class Resolved implements Serializable, Writeable {
```

Comment:
Why is Serializable, Writeable dropped ?

Reply:
`Resolved` is never passed over the wire, so it does not need to be `Serializable` and/or `Writeable`.